src/System.Private.ServiceModel/tools/PRService/pr.ashx

```diff
@@ -0,0 +1,217 @@
+{
+    private const string _gitExecutablePath = @"C:\Program Files\Git\bin\git.exe";
+    private static string _gitRepoPath;
+    private const int _gitExecutionTimeoutMilliseconds = 100000;
+
+    public void ProcessRequest(HttpContext context)
+    {
+        _gitRepoPath = context.Request.PhysicalApplicationPath + "..\\..\\..\\..";
+        string prString = context.Request.QueryString["pr"];
+
+        context.Response.Write(string.Format("PR = {0}<br/>", HttpUtility.HtmlEncode(prString) ?? "Unspecified"));
+
+        StringBuilder result = new StringBuilder();
+
+        bool success = CheckPrerequisites(result);
+
+        if (success) 
+        {
+            success = CleanupMergedBranches(result); 
+        }
+
+        if (success)
+        {
+            // Check inputs to the script
+            uint pr;
+            if (!string.IsNullOrEmpty(prString) && uint.TryParse(prString, out pr))
+            {
+                success = SyncToPr(pr, result);
+            }
+            else
+            {
+                context.Response.Write(string.Format("Invalid PR ID specified<br/>"));
+                success = false;
+            }
+        } 
+        
+        if (!success)
+        {
+            context.Response.StatusCode = 400;
+        }
+
+        if (context.Request.IsLocal)
+        {
+            // Don't allow error results to leak to non-local machines
+            context.Response.Write(HttpUtility.HtmlEncode(result));
+        }
+    }
+
+    // Check prerequisites for whether or not we can run
+    private bool CheckPrerequisites(StringBuilder executionResult)
+    {
+        if (!File.Exists(_gitExecutablePath))
+        {
+            executionResult.AppendFormat("Could not find or insufficient permissions to access the git executable at path '{0}'", _gitExecutablePath);
+            return false;
+        }
+
+        if (!Directory.Exists(_gitRepoPath))
+        {
+            executionResult.AppendFormat("Could not find or insufficient permissions to access the git repo at path '{0}'", _gitRepoPath);
+            return false;
+        }
+
+        // Do a sanity execute of the git executable 
+
+        try
+        {
+            string[] gitCommands = new string[]
+            {
+                "status"
+            };
+
+            RunGitCommands(gitCommands, executionResult);
+        }
+        catch (Exception ex)
+        {
+            executionResult.AppendFormat("Exception caught while attempting to start '{0}'", _gitExecutablePath);
+            executionResult.Append(ex.ToString());
+            return false;
+        }
+
+        return true;
+    }
+
+    private bool SyncToPr(uint pr, StringBuilder executionResult)
+    {
+        string[] gitCommands = new string[]
+        {
+            "clean -fdx",
+            "checkout master",
+            "fetch --all --prune",
+            "merge --ff-only origin/master",
+            string.Format("fetch origin pull/{0}/head:pr/{0}", pr),
```

Comment:
Something to consider in future -- technically any failure here should be a 500 -- except for the fetch of the PR, which would likely be due to a bogus PR#.  I know, I can hear the argument we won't call this service with a bogus PR, but if we go there and say requests are always well formed and correct, then 400 would never be a correct response.